theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,209 @@
+    }
+
+    npy_int minibatch_size = PyArray_DIMS( in_activations )[1];
+    npy_int alphabet_size = PyArray_DIMS( in_activations )[2];
+
+    void * ctc_cpu_workspace = NULL;
+
+    npy_float32 * costs = NULL;
+    npy_intp cost_size = minibatch_size;
+
+    if ( NULL == (*out_costs) )
+    {
+        // Symbolic variable has no memory backing, so we create one
+        *out_costs = (PyArrayObject *) PyArray_ZEROS( 1, &cost_size, NPY_FLOAT32, 0 );
+    }
+    else if ( PyArray_NDIM( *out_costs ) != 1 ||
+              PyArray_DIMS( *out_costs )[0] != cost_size )  // matrix has the wrong size
+    {
+        Py_XDECREF( *out_costs ); 
+        // Allocate new matrix
+        *out_costs = (PyArrayObject *) PyArray_ZEROS( 1, &cost_size, NPY_FLOAT32, 0 );
+    }
+
+    if ( NULL == (*out_costs) )
+    {
+        PyErr_Format( PyExc_ValueError,
+                      "Could not allocate storage for CTC costs" );
+        return 1;
+    }
+    
+    costs = (npy_float32 *) PyArray_DATA( *out_costs );
+
+    if ( NULL == (*out_gradients) )
+    {
+        // Symbolic variable has no real backing, so create one.
+        *out_gradients = (PyArrayObject*) PyArray_ZEROS( 3, PyArray_DIMS( in_activations ),
+            NPY_FLOAT32, 0 );
+    }
+    else if ( PyArray_NDIM( *out_gradients ) != 3 
+        || PyArray_DIMS( *out_gradients )[0] != PyArray_DIMS( in_activations )[0]
+        || PyArray_DIMS( *out_gradients )[1] != PyArray_DIMS( in_activations )[1]
+        || PyArray_DIMS( *out_gradients )[2] != PyArray_DIMS( in_activations )[2] )
+    {
+        // Existing matrix is the wrong size. Make a new one.
+        // Decrement ref counter to existing array
+        Py_XDECREF( *out_gradients ); 
+        // Allocate new array
+        *out_gradients = (PyArrayObject *) PyArray_ZEROS(3, PyArray_DIMS( in_activations ),
+            NPY_FLOAT32, 0);
+    }
+
+    if ( NULL == (*out_gradients) )
+    {
+        PyErr_Format( PyExc_ValueError,
+                      "Could not allocate storage for CTC gradients!" );
+        return 1;        
+    }
+
+    npy_float32 * gradients = (npy_float32 *) PyArray_DATA( *out_gradients );
+
+    ctcStatus_t status;
+    size_t cpu_workspace_size;
+
+    status = get_workspace_size( label_lengths, input_lengths, alphabet_size,
+        minibatch_size, ctc_options, &cpu_workspace_size );
+
+    if ( CTC_STATUS_SUCCESS != status )
+    {
+        PyErr_Format( PyExc_ValueError,
+                      "Could not compute the CTC workspace size!" );
+        return 1;    
+    }
+
+    ctc_cpu_workspace = malloc( cpu_workspace_size );
+
+
+    status = compute_ctc_loss( activations, gradients, flat_labels,
+        label_lengths, input_lengths, alphabet_size, minibatch_size,
+        costs, ctc_cpu_workspace, ctc_options );
+
+    if ( CTC_STATUS_SUCCESS != status )
+    {
+        PyErr_Format( PyExc_ValueError, "Failed to compute CTC loss!" );
+        return 1;         
+    }
+
+    if ( NULL != activations_copy )
```

Comment:
If an error occurs above, the function will return without deallocating `activation_copy` if needed.

In general, make sure that, at every point where an error occurs, any allocated memory is deallocated before the function returns!